src/main/java/de/qabel/desktop/storage/BoxVolume.java

```diff
@@ -0,0 +1,113 @@
+package de.qabel.desktop.storage;
+
+import com.amazonaws.auth.AWSCredentials;
+import de.qabel.core.crypto.CryptoUtils;
+import de.qabel.core.crypto.DecryptedPlaintext;
+import de.qabel.core.crypto.QblECKeyPair;
+import de.qabel.desktop.exceptions.QblStorageException;
+import org.apache.commons.io.IOUtils;
+import org.spongycastle.crypto.InvalidCipherTextException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.UUID;
+
+public class BoxVolume {
+
+	private static final Logger logger = LoggerFactory.getLogger(BoxVolume.class.getName());
+
+	StorageReadBackend readBackend;
+	StorageWriteBackend writeBackend;
+
+	private QblECKeyPair keyPair;
+	private byte[] deviceId;
+	private CryptoUtils cryptoUtils;
+	private File tempDir;
+
+	public BoxVolume(String bucket, String prefix, AWSCredentials credentials,
+	                 QblECKeyPair keyPair, byte[] deviceId, File tempDir) {
+		this(new S3ReadBackend(bucket, prefix), new S3WriteBackend(credentials, bucket, prefix),
+				keyPair, deviceId, tempDir);
+	}
+
+	public BoxVolume(StorageReadBackend readBackend, StorageWriteBackend writeBackend,
+	                 QblECKeyPair keyPair, byte[] deviceId, File tempDir) {
+		this.keyPair = keyPair;
+		this.deviceId = deviceId;
+		this.readBackend = readBackend;
+		this.writeBackend = writeBackend;
+		cryptoUtils = new CryptoUtils();
+		this.tempDir = tempDir;
+		try {
+			loadDriver();
+		} catch (ClassNotFoundException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected void loadDriver() throws ClassNotFoundException {
+		logger.info("Loading PC sqlite driver");
+		Class.forName("org.sqlite.JDBC");
+	}
+
+	public BoxNavigation navigate() throws QblStorageException {
+		String rootRef = getRootRef();
+		logger.info("Navigating to " + rootRef);
+		InputStream indexDl = readBackend.download(rootRef);
+		File tmp;
+		try {
+			byte[] encrypted = IOUtils.toByteArray(indexDl);
+			DecryptedPlaintext plaintext = cryptoUtils.readBox(keyPair, encrypted);
+			// Should work fine for the small metafiles
```

Comment:
Big metafiles are larger than the maximum size for a byte array on the system. "Small" is small enough that handling the whole file in a byte array doesn't blow up.
The problem is that the whole file is read into memory because the CryptoUtils noise box methods only accept byte arrays.

I guess the threshold size is reached  with about round about 50k files in a single folder. I don't think this problem will become relevant because the file system does not scale to that many files in a single folder because they are all saved in a single SQLite database.